wardrobe/timestamp/src/lib.rs

```diff
@@ -0,0 +1,382 @@
+    TooManyOutputsWhileSettingTimestamp,
+    /// The previous timestamp that is peeked at must be from the immediate ancestor block, but this one is not.
+    PreviousTimestampWrongHeight,
+    /// No previous timestamp was peeked at in this transaction, but at least one peek is required.
+    MissingPreviousTimestamp,
+    /// Inputs were specified while setting the timestamp, but none are allowed.
+    InputsWhileSettingTimestamp,
+    /// The new timestamp is not sufficiently far after the previous (or may even be before it).
+    TimestampTooOld,
+
+    /// When cleaning up old timestamps, you must supply exactly one peek input which is the "new time reference"
+    /// All the timestamps that will be cleaned up must be at least the CLEANUP_AGE older than this reference.
+    CleanupRequiresOneReference,
+    /// When cleaning up old timestamps, you may not create any new state at all.
+    /// However, you have supplied some new outputs in this transaction.
+    CleanupCannotCreateState,
+    /// You may not clean up old timestamps until they are at least the CLEANUP_AGE older than another
+    /// noted timestamp on-chain.
+    DontBeSoHasty,
+}
+
+/// A constraint checker for the simple act of setting a new best timetamp.
+///
+/// This is expected to be performed through an inherent, and to happen exactly once per block.
+///
+/// This transaction comsumes a single input which is the previous best timestamp,
+/// And it creates two new outputs. A best timestamp, and a noted timestamp, both of which
+/// include the same timestamp. The purpose of the best timestamp is to be consumed immediately
+/// in the next block and guarantees that the timestamp is always increasing by at least the minimum.
+/// On the other hand, the noted timestamps stick around in storage for a while so that other
+/// transactions that need to peek at them are not immediately invalidated. Noted timestamps
+/// can be voluntarily cleand up later by another transaction.
+#[cfg_attr(feature = "std", derive(Serialize, Deserialize))]
+#[derive(Encode, Decode, DebugNoBound, DefaultNoBound, PartialEq, Eq, CloneNoBound, TypeInfo)]
+#[scale_info(skip_type_params(T))]
+pub struct SetTimestamp<T>(PhantomData<T>);
+
+impl<T: TimestampConfig + 'static, V: Verifier + From<UpForGrabs>> ConstraintChecker<V>
+    for SetTimestamp<T>
+{
+    type Error = TimestampError;
+    type InherentHooks = TuxedoInherentAdapter<Self>;
+
+    fn check(
+        &self,
+        input_data: &[tuxedo_core::types::Output<V>],
+        peek_data: &[tuxedo_core::types::Output<V>],
+        output_data: &[tuxedo_core::types::Output<V>],
+    ) -> Result<TransactionPriority, Self::Error> {
+        log::debug!(
+            target: LOG_TARGET,
+            "🕰️🖴 Checking constraints for SetTimestamp."
+        );
+
+        // FRAME pallet authors are required to ensure_none! to make sure this is an inherent.
```

Comment:
This comment might not make so much sense if the person using Tuxedo has never used FRAME but I dont mind it all that much

Reply:
730028a0ee7398213d3b79abdca209ff5356cb34